docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+
+For example the record literal `{a:1, b:2}` is just shorthand for `{a:1 | {b:2 | {} }}` whose type is given by `{a:int | {b:int | {} }}`.
+
+More generally, the polymorphic type of the extension operation `{_ with l:_}` is given by:
+
+```
+forall ['a, 'r] where 'r: Rec ('a, 'r) -> {l: 'a | 'r}
+```
+
+This type is read as, "for all type variables _'a_ and _'r_ where _'r_ is a record kind, we have a function type taking _'a_ and _'r_ as input and returning a record type extended from _'r_ by adding a field _l_ of type _'a_".
+Selection and restriction operations are defined similarly, and update and rename can be expressed using these basic operations.
+
+Unification of record types is particularly elegant and can be expressed using the following 3 rules:
+
+1. `{a: 't | 'r} = {a: 'u | 's} => 't = 'u and 'r = 's`
+2. `{a: 't | 'r} = {b: 'u | 's} => {a: 't | 'v} = 's and {b: 'u | 'v} = 'r`
+3. `{a: 't | 'r} = {b: 'u | 'r} => error`
+
+Note here `=` is synonymous for unify.
+
+### Implications
+
+The Flux type system extends Hindley-Milner with a _kind_ system.
+Currently records are modeled as _kinds_ in this system.
+Extensible records however fall entirely within the _type_ system, not the _kind_ system.
+Describing records this way allows us to break the mutual recursion between type unification and kind unification.
+Proving the correctness of type inference now follows directly from Hindley-Milner.
+
+### Caveats
+
+* Nulls
+
+    Extensible records do not support accessing non-existent fields on an record.
+    Such operations will result in a unification error.
+    The Flux language specification states that accessing a non-existent field on a record will return an untyped null value.
+    The spec will need to be updated accordingly as this will no longer be the case.
+
+    Null values will still be possible at runtime but they must be typed nulls.
+    The concept of an untyped null will need to be removed from the spec.
+
+* Polymorphic Record Literals
+
+    The current record _kinds_ allow for polymorphic record literals.
+    For example, currently you can define a list of heterogeneous records like so:
+    ```
+    a = [ {a:1, b:"str"}, {b:"str", c:3.3} ]
+    ```
+
+    This would result in a unification error in the case of extensible records.
+    What this means is that we lose some of our ability to represent polymorphic container types.
+
+    This might be concerning since the 2nd goal of this iteration is to represent table streams as a list of **heterogeneous** record types.
+    However such concern is unwarranted as streams cannot be constructed from atomic types the way list literals can.
+    Instead they are obtained entirely via a side effect operation - a call to the database.
+    This means that from the point of view of the type system, a table stream is just a list of **homogeneous** records.
+    In other words, record fields or columns that are never accessed never need to be typed.
```

Comment:
Is this the same as saying that we now only track the lower bound of records? And the upper bound is now implicitly the set of all labels? (Using the lower, upper bound concept of kinded records)

Reply:
Yes. For example, take a record type like `{a:int | 'r }`. In the literature `'r` is referred to as a row variable. Technically `'r` is a type variable representing some unknown record type. Intuitively we can think of `'r` as a variable that ranges over every possible field or label.